### PR TITLE
[MRG+1] Dns cache size and timeout options

### PR DESCRIPTION
**Pull request by sibiryakov**
Continuing of #1092 and #1120.

### CODE REVIEW

**Comment by kmike** [comment]
Looks good!

**Comment by nyov** [comment]
I'm sorry if you've read this, but I'll repost in case my comment got lost.

The `getHostByName` interface in theory supports a list of values (1, 3, 11, 45), which contains staggerable timeouts. But it is only `sum()`-med in the default implementation.

To quote [the docs](https://twistedmatrix.com/documents/14.0.1/api/twisted.internet.interfaces.IResolverSimple.html#getHostByName): "timeout: Number of seconds after which to reissue the query. When the last timeout expires, the query is considered failed. (type: Sequence of int)"

There is no support for float values from the interface point of view, but support for a sequence.

**Comment by sibiryakov** [comment]
Unfortunately sequence doesn't work in ThreadedResolver. Instead it uses IReactorTime.callLater which has seconds argument as float. See http://twistedmatrix.com/documents/current/api/twisted.internet.interfaces.IReactorTime.html#callLater 
Previously we decided, that it doesn't make much sense to introduce interface changes that doesn't work. IMO, the best would be to wait until twisted.names resolver will be fixed and then perform necessary changes, e.g. switch to sequence of ints.

**Comment by nyov** [comment]
> IMO, the best would be to wait until twisted.names resolver will be fixed and then perform necessary changes, e.g. switch to sequence of ints.

But then it would be an avoidable change to a user-facing setting, moving DNS_TIMEOUT from a float argument to a sequence later?

**Comment by sibiryakov** [comment]
We could support both then. By checking the type of value in settings, and using one timeout or sequence.

**Comment by nyov** [comment]
Never mind, if you have discussed this and consider it good, lets get some of these PRs merged.
This release keeps dragging on :tired_face:

**Comment by kmike** [comment]
@nyov sorry, we discussed a list vs a single value in a private chat. The problem with list is that from Scrapy user point of view it is meaningless - "provide a list and we'll sum up the values" is a strange API. As @sibiryakov said, we can always add another setting in future, or just allow a list of values as well as a single value for this option. 

This PR (like many others) is not marked as [Scrapy 1.0](https://github.com/scrapy/scrapy/milestones/Scrapy%201.0), so merging it is not a requirement for 1.0 release. It just happens that when you start paying more attention to PRs more PRs begin to appear :)

**Comment by nyov** [comment]
> It just happens that when you start paying more attention to PRs more PRs begin to appear :)

Aye. And I was hoping to smuggle some more in as well :] So it's partially my own fault. But as most of the still open tickets for 1.0 are things I can't really contribute to, there is nothing else for me but finding other stuff to push.

**Comment by sibiryakov** [comment]
Ping! Guys, what about merging it?

File: scrapy/resolver.py [hunk 1]
```diff
@@ -4,15 +4,20 @@
 from scrapy.utils.datatypes import LocalCache
 
 # TODO: cache misses
-# TODO: make cache size a setting
 
 dnscache = LocalCache(10000)
```
Comment (nyov) [inline scrapy/resolver.py]: Shouldn't this limit change with `DNSCACHE_SIZE`? Or is this the upper hard limit.

**Pull request merged**